torch/_higher_order_ops/cond.py

```diff
@@ -404,3 +410,50 @@ def cond_func(ctx, pred, true_fn, false_fn, inputs):
             unwrapped_pred, functional_true, functional_false, unwrapped_inputs
         )
         return ctx.wrap_tensors(cond_return)
+
+
+@cond_op.py_impl(torch._C._functorch.TransformType.Vmap)
+def cond_batch_rule(interpreter, pred, true_fn, false_fn, inputs):
+    assert isinstance(
+        inputs, (list, tuple)
+    ), "Cond inputs must be a list or tuple of tensors"
+    assert all(
+        isinstance(i, torch.Tensor) for i in inputs
+    ), "Cond inputs must be a list of tensors"
+
+    pred_ = get_unwrapped(pred) if is_batchedtensor(pred) else pred
+
+    # unbatched tensors are not vmapped
+    tensors, in_dims = zip(
+        *[
+            (get_unwrapped(t), maybe_get_bdim(t)) if is_batchedtensor(t) else (t, None)
+            for t in inputs
+        ]
+    )
+
+    if is_batchedtensor(pred):
+        # prepend "pred" and vmap everything
+        tensors = (pred_,) + tensors
+        in_dims = (0,) + in_dims
+
+        def fn(p, *args):
+            t = true_fn(*args)
+            f = false_fn(*args)
+            return torch.where(p, t[0], f[0])
+
+        with interpreter.lower():
+            result = torch.vmap(fn, in_dims=in_dims)(*tensors)
+
+    else:
+        # predicate is known at this stage and it is a boolean expression or a
+        # tensor with one element.
+        true_fn = torch.vmap(true_fn, in_dims=in_dims)
+        false_fn = torch.vmap(false_fn, in_dims=in_dims)
```

Comment:
Should probably happen in a follow-up PR, but for performance reasons, we should call something like resume_vmap here instead of torch.vmap (https://github.com/pytorch/pytorch/blob/main/torch/_functorch/vmap.py#L418).

The problem is: torch.vmap always returns a Tensor with the batch dimension, even if it didn't need it. In the following case:
```
torch.cond(pred, lambda x: torch.ones(3), lambda x: torch.zeros(3), (x,)
```

if we do
```py
x = torch.randn(2)
y = vmap(lambda x: torch.ones(3))(x)
```
We get that y is a tensor of shape (2, 3).  To make things more efficient, we want to return a Tensor with shape (3,).